### PR TITLE
Variant 4 Tweak 2 [Merge other Variant4-First]

### PR DESCRIPTION
**Pull request by vtnerd**
This is a second tweak for variant-4 cryptonote:

* The values used in the first shuffle operation are now used to modify the selection of the second memory address.
* The values used in the second shuffle operation are now used to modify the input to the next AES operation (next loop iteration).
* The output (destination registers) of the random program are used in the writeback value of the second memory address and the memory address calculation for the next AES operation (next loop iteration).
* Tweak2-2 was dropped.

The operations done are fairly simple, but AFAIK cannot be optimized out. They enforce a stronger order of operations. Feedback is welcome. This is based on @SChernykh PR, only the last commit is relevant to this PR.

### CODE REVIEW

**Comment by SChernykh** [comment]
These changes look good at the first glance, but they're bigger than I expected. It's not bad, but it'll take me some time to implement it in miners. I'll first try it in xmrig for CPU and xmrig-amd and see how it affects performance.

**Comment by SChernykh** [comment]
> * The output (destination registers) of the random program are used in the writeback value of the second memory address and the memory address calculation for the next AES operation (next loop iteration).

This will be bad for performance. Random math result is not available for the whole main loop iteration, it runs in parallel with the rest of main loop. If you use registers R0-R4 anywhere else, you'll likely stall the loop. Unless it's only written to memory and doesn't have additional dependencies. I'll check how it behaves in practice.

**Comment by SChernykh** [comment]
> * The output (destination registers) of the random program are used in the writeback value of the second memory address and the memory address calculation for the next AES operation (next loop iteration).

I'll start implementing it with and without this one and compare performance on CPU and GPU. Then we can decide what to do next.

**Comment by SChernykh** [comment]
@vtnerd @moneromooo-monero 
I've tested it on Ryzen 5 2600 @ 4 GHz: 
- 88.2-88.6 H/s with vtnerd's tweaks 
- 91.6-93.3 H/s without vtnerd's tweaks.

It's only 4-5% slower. Note that it was unoptimized code that I quickly hacked together. I think that there is still room for 1-2% performance improvement on CPU, so these tweaks won't change CPU performance much. GPU performance shouldn't change too.

I'll implement it in xmrig/xmrig-amd/xmrig-nvidia and run more tests on CPU/GPU. Chances are good that no further changes will be required.

File: src/crypto/slow-hash.c [hunk 1]
```diff
@@ -116,48 +117,74 @@ extern void aesb_pseudo_round(const uint8_t *in, uint8_t *out, const uint8_t *ex
...
+    uint64x2_t chunk1 = vld1q_u64(U64((base_ptr) + ((offset) ^ 0x10))); \
     const uint64x2_t chunk2 = vld1q_u64(U64((base_ptr) + ((offset) ^ 0x20))); \
     const uint64x2_t chunk3 = vld1q_u64(U64((base_ptr) + ((offset) ^ 0x30))); \
     vst1q_u64(U64((base_ptr) + ((offset) ^ 0x10)), vaddq_u64(chunk3, vreinterpretq_u64_u8(_b1))); \
     vst1q_u64(U64((base_ptr) + ((offset) ^ 0x20)), vaddq_u64(chunk1, vreinterpretq_u64_u8(_b))); \
     vst1q_u64(U64((base_ptr) + ((offset) ^ 0x30)), vaddq_u64(chunk2, vreinterpretq_u64_u8(_a))); \
+    if (variant >= 4) \
+    { \
+      chunk1 = veorq_u64(chunk1, chunk2); \
+      _c = veorq_u64(_c, chunk3); \
```
Comment (SChernykh) [inline src/crypto/slow-hash.c]: Compilation error on ARM:
```
/opt/buildbot/builder/monero-linux-armv8/build/src/crypto/slow-hash.c:146:22: error: incompatible type for argument 1 of ‘veorq_u64’
       _c = veorq_u64(_c, chunk3); \
                      ^
```

File: src/crypto/variant4_random_math.h [hunk 2]
```diff
@@ -0,0 +1,441 @@
...
+	const int asic_op_latency[V4_INSTRUCTION_COUNT] = { 3, 1, 1, 1, 1, 1 };
+
+	// Available ALUs for each instruction
+	const int op_ALUs[V4_INSTRUCTION_COUNT] = { ALU_COUNT_MUL, ALU_COUNT, ALU_COUNT, ALU_COUNT, ALU_COUNT, ALU_COUNT };
+
+	int8_t data[32];
+	memset(data, 0, sizeof(data));
+	uint64_t tmp = SWAP64LE(height);
+	memcpy(data, &tmp, sizeof(uint64_t));
+	data[20] = 0xda; // change seed
```
Comment (SChernykh) [inline src/crypto/variant4_random_math.h]: 0xda is greater than 127, it doesn't fit into int8_t range of values. It's better to write
```
data[20] = -38;
```

**Comment by moneromooo-monero** [comment]
This ended up merged within https://github.com/monero-project/monero/pull/5126

+duplicate

**Pull request closed without merging**